runtime/src/chplsys.c

```diff
@@ -299,40 +306,60 @@ static void computeHeapPageSizeByGuessing(size_t page_size_in)
   }
 }
 
-void chpl_computeHeapPageSize(void) {
-  size_t pageSize = 0;
-
+static void computeHeapPageSize(void) {
 #if defined __linux__
-  char* ev;
-  if ((ev = getenv("HUGETLB_DEFAULT_PAGE_SIZE")) == NULL)
-    pageSize = chpl_getSysPageSize();
-  else {
+  //
+  // If we're using hugepages explicitly, just go with what it says.
+  // Otherwise, try to figure it out ourselves.
+  //
+  {
+    char* ev;
+
+    if ((ev = getenv("HUGETLB_DEFAULT_PAGE_SIZE")) != NULL) {
+      int  num_scanned;
+      char units;
 
-    size_t tmpPageSize;
-    int  num_scanned;
-    char units;
+      if ((num_scanned = sscanf(ev, "%zi%c", &heapPageSize, &units)) != 1) {
+        if (num_scanned != 2 || strchr("kKmMgG", units) == NULL) {
+          chpl_internal_error("unexpected HUGETLB_DEFAULT_PAGE_SIZE syntax");
+        }
 
-    if ((num_scanned = sscanf(ev, "%zi%c", &tmpPageSize, &units)) != 1) {
-      if (num_scanned == 2 && strchr("kKmMgG", units) != NULL) {
         switch (units) {
-        case 'k': case 'K': tmpPageSize <<= 10; break;
-        case 'm': case 'M': tmpPageSize <<= 20; break;
-        case 'g': case 'G': tmpPageSize <<= 30; break;
+        case 'k': case 'K': heapPageSize <<= 10; break;
+        case 'm': case 'M': heapPageSize <<= 20; break;
+        case 'g': case 'G': heapPageSize <<= 30; break;
         }
       }
-      else {
-        chpl_internal_error("unexpected HUGETLB_DEFAULT_PAGE_SIZE syntax");
-      }
+
+      return;
     }
+  }
+#endif
 
-    pageSize = tmpPageSize;
+  // note: sets heapPageSize
+  computeHeapPageSizeByGuessing(chpl_getSysPageSize());
+}
+
+
+size_t chpl_getHeapPageSize(void) {
+#ifdef _POSIX_VERSION
+  {
+    static pthread_once_t onceCtl = PTHREAD_ONCE_INIT;
+
+    if (pthread_once(&onceCtl, computeHeapPageSize) != 0) {
+      chpl_internal_error("pthread_once() failed");
+    }
   }
 #else
-  pageSize = chpl_getSysPageSize();
+  //
+  // Note: no single-threaded initialization for non-POSIX OS.
+  //
```

Comment:
I wonder if we should print a warning in this case? Or use an atomic or something? It seems if we end up in this case, it could have disasterous performance consequences.